### PR TITLE
Get `totalResults` Value by Total User Count in SCIM2 User API

### PR DESCRIPTION
**Pull request by dhaura**
### Purpose
- When the SCIM API is invoked with a filter, the current behavior involves returning a predefined value for the totalResults parameter based on the MaxUserNameListLength, rather than reflecting the actual results that match the filter criteria. This fix addresses the issue by accurately determining the user count that corresponds to the given filter. 
- Furthermore, when invoking user endpoint with groups filter by passing an role (hybrid or system roles), still the endpoint returns a list of userd which a linked to the passed role. But when role group seperation is enabled, it should return an empty list since the passed value is for a role (not for a group). This is also fixed with this PR.
- To enable these fixes, the following config should be added.
```
[scim2]
enable_group_based_user_filter_improvements = true
```

### Related Issues
- https://github.com/wso2/product-is/issues/12813

### Related PRs
- https://github.com/wso2/carbon-kernel/pull/3906
- https://github.com/wso2/carbon-kernel/pull/3951
- https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/533

### CODE REVIEW

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java [hunk 1]
```diff
@@ -1457,6 +1462,29 @@ private UsersGetResponse filterUsersBySingleAttribute(ExpressionNode node, Map<S
...
+     * method to get user count by filtering parameter.
+     *
+     * @param node       Expression node for single attribute filtering
+     * @param offset     Starting index of the count
+     * @param limit      Counting value
+     * @param sortBy     SortBy
+     * @param sortOrder  Sorting order
+     * @param domainName Domain to run the filter
+     * @return User count
+     * @throws BadRequestException
```
Comment (dhaura) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java:1475]: Describe exceptions in all places

Comment (dhaura) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java]: Resolved with https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/547/commits/6971d7032dd66875fb60641349cbae8585474a28

**Comment by mpmadhavig** [comment]
@dhaura could you please list out all the flows you covered while testing.

Also, if you haven't done already, test this with count query param added in the request.
- count > 100 -> returned count 100 and totalResults should give the actual amount.
- count < 100 -> returned count should be the specified amount and totalResults should give the actual amount.

**Comment by dhaura** [comment]
@mpmadhavig I tested this fix with the SCIM2 user endpoint groups filter using API call [1] for following scenarios.

- Groups created by organization admin. (ex: `Stundet`)
    - For 10 and 110 users per group.
- Hybrid Roles: `Administrator`, `everyone`


[1] - 
```
curl --location 'https://<IS_DOMAIN:PORT>/t/<TENANT_DOMAIN>/scim2/Users?filter=groups%2Beq%2BStudent' \
--header 'Authorization: Bearer <TOKEN>'
```
> Also, if you haven't done already, test this with count query param added in the request.
> - count > 100 -> returned count 100 and totalResults should give the actual amount.

- count = 100
Screenshot 2024-04-04 at 09 35 34

- count = 105
Screenshot 2024-04-04 at 09 28 47

> - count < 100 -> returned count should be the specified amount and totalResults should give the actual amount.

- count = 10
Screenshot 2024-04-04 at 09 28 28

- count = 50
Screenshot 2024-04-04 at 09 28 36

**Review by ashensw — APPROVED** [review]

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java [hunk 2]
```diff
@@ -1446,7 +1446,14 @@ private UsersGetResponse filterUsersBySingleAttribute(ExpressionNode node, Map<S
                     maxLimit = Math.max(maxLimit, limit);
                 }
                 // Get total users based on the filter query without depending on pagination params.
-                totalResults += filterUsers(node, 1, maxLimit, sortBy, sortOrder, domainName).size();
+                if (SCIMCommonUtils.isGroupBasedUserFilteringImprovementsEnabled() &&
+                        (isJDBCUSerStore(domainName) || isAllConfiguredUserStoresJDBC())) {
+                    // Get the total user count by the filter query.
+                    // This is only implemented for JDBC userstores.
```
Comment (ShanChathusanda93) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java:1452]: Can we make this a single line comment? Or else format as a multi line comment?

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java [hunk 3]
```diff
@@ -1457,6 +1464,30 @@ private UsersGetResponse filterUsersBySingleAttribute(ExpressionNode node, Map<S
         return getDetailedUsers(filteredUsers, totalResults);
     }
 
+    /**
+     * method to get user count by filtering parameter.
```
Comment (ShanChathusanda93) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java:1468]: Suggestion : `Get user count by filtering parameter.`

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java [hunk 4]
```diff
@@ -1694,6 +1730,54 @@ private Set<org.wso2.carbon.user.core.common.User> filterUsers(Node node, int of
         }
     }
 
+    /**
+     * Method to get User Count by Group filter
```
Comment (ShanChathusanda93) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java:1734]: Suggestion : `Get user count by group filter`

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java [hunk 5]
```diff
@@ -1694,6 +1730,54 @@ private Set<org.wso2.carbon.user.core.common.User> filterUsers(Node node, int of
...
+        // Set filter values.
+        String attributeName = ((ExpressionNode) node).getAttributeValue();
+        String filterOperation = ((ExpressionNode) node).getOperation();
+        String attributeValue = ((ExpressionNode) node).getValue();
+
+        /*
+        If there is a domain and if the domain separator is not found in the attribute value, append the domain
+        with the domain separator in front of the new attribute value.
+        */
+        attributeValue = UserCoreUtil.addDomainToName(((ExpressionNode) node).getValue(), domainName);
```
Comment (ShanChathusanda93) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java:1755]: Do we need the extra assignment in 1749?

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java (hunk 2 above)
Comment (dhaura) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java:1452]: Addressed with https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/553

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java (hunk 3 above)
Comment (dhaura) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java:1468]: Addressed with https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/553

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java (hunk 4 above)
Comment (dhaura) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java:1734]: Addressed with https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/553

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java (hunk 5 above)
Comment (dhaura) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java:1755]: Addressed with https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/553

**Pull request merged**